File: xaga/tools/sleepy/src/utils/traits.rs
// Copyright 2026 rainy-juzixiao
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
use clang::Entity;

pub fn is_noexcept(entity: &Entity) -> bool {
    entity.get_exception_specification()
        .map(|s| matches!(s,
            clang::ExceptionSpecification::BasicNoexcept |
            clang::ExceptionSpecification::NoThrow
        ))
        .unwrap_or(false)
}

pub fn is_override(entity: &Entity) -> bool {
    // libclang 没有直接 API，但 CXCursor_CXXOverrideAttr 子节点能判断
    entity.get_children().iter().any(|c| {
        format!("{:?}", c.get_kind()).contains("Override")
    })
}

pub fn is_deleted(entity: &Entity) -> bool {
    entity.get_availability() == clang::Availability::Unavailable
}

pub fn is_defaulted(entity: &Entity) -> bool {
    entity.is_defaulted()
}

pub fn is_constexpr(entity: &Entity) -> bool {
    // clang crate 暂无直接 API，通过 children 里找 ConstexprAttr
    entity.get_children().iter().any(|c| {
        format!("{:?}", c.get_kind()).contains("Constexpr")
    })
}

pub fn is_inline(entity: &Entity) -> bool {
    entity.is_inline_function()
}

// is_ref_lvalue / is_ref_rvalue 用 display_name 里的签名解析
pub fn is_ref_lvalue(entity: &Entity) -> bool {
    entity.get_display_name()
        .map(|s| s.ends_with(" &") || s.contains(") & ") || s.contains(") &\n"))
        .unwrap_or(false)
}

pub fn is_ref_rvalue(entity: &Entity) -> bool {
    entity.get_display_name()
        .map(|s| s.ends_with(" &&") || s.contains(") && ") || s.contains(") &&\n"))
        .unwrap_or(false)
}

pub fn is_explicit(entity: &Entity) -> bool {
    entity.get_children().iter().any(|c| {
        format!("{:?}", c.get_kind()).contains("Explicit")
    })
}

pub fn is_final(entity: &Entity) -> bool {
    entity.get_children().iter().any(|c| {
        format!("{:?}", c.get_kind()).contains("Final")
    })
}

pub fn is_consteval(entity: &Entity) -> bool {
    entity.get_children().iter().any(|c| {
        format!("{:?}", c.get_kind()).contains("Consteval")
    })
}

File: tektite/library/core/tests/typeinfo/traits.cc
/*
* Copyright 2026 rainy-juzixiao
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <catch2/catch_test_macros.hpp>
#include <rainy/core/typeinfo.hpp>

using namespace rainy::core;

struct TestClass
{
    int value;
};

enum class TestEnum
{
    A,
    B
};

union TestUnion
{
    int i;
    float f;
};

TEST_CASE("fundamental traits")
{
    STATIC_REQUIRE(typeinfo::create<int>().is_integer());
    STATIC_REQUIRE(typeinfo::create<int>().is_arithmetic());

    STATIC_REQUIRE(typeinfo::create<float>().is_floating_point());
    STATIC_REQUIRE(typeinfo::create<float>().is_arithmetic());
}

TEST_CASE("pointer traits")
{
    STATIC_REQUIRE(typeinfo::create<int*>().is_pointer());
}

TEST_CASE("reference traits")
{
    STATIC_REQUIRE(typeinfo::create<int&>().is_reference());
    STATIC_REQUIRE(typeinfo::create<int&>().is_lvalue_reference());
    STATIC_REQUIRE(typeinfo::create<int&&>().is_rvalue_reference());
}

TEST_CASE("cv traits")
{
    STATIC_REQUIRE(typeinfo::create<const int>().is_const());
    STATIC_REQUIRE(typeinfo::create<volatile int>().is_volatile());
    STATIC_REQUIRE(typeinfo::create<const volatile int>().is_const_volatile());
}

TEST_CASE("class traits")
{
    STATIC_REQUIRE(typeinfo::create<TestClass>().is_class());
}

TEST_CASE("enum union traits")
{
    STATIC_REQUIRE(typeinfo::create<TestEnum>().has_traits(traits::is_enum));
    STATIC_REQUIRE(typeinfo::create<TestUnion>().has_traits(traits::is_union));
}